ludwig/backend/ray.py

```diff
@@ -0,0 +1,174 @@
+        num_hosts=len(best_resources),
+        use_gpu=use_gpu
+    )
+
+
+class RayRemoteModel:
+    def __init__(self, model):
+        self.cls, self.args, state = list(model.__reduce__())
+        self.state = ray.put(state)
+
+    def load(self):
+        obj = self.cls(*self.args)
+        obj.__setstate__(ray.get(self.state))
+        return obj
+
+
+class RayTrainer(BaseTrainer):
+    def __init__(self, horovod_kwargs, trainer_kwargs):
+        setting = RayExecutor.create_settings(timeout_s=30)
+        self.executor = RayExecutor(setting, **{**get_horovod_kwargs(), **horovod_kwargs})
+        self.executor.start(executable_cls=RemoteTrainer, executable_kwargs=trainer_kwargs)
+
+    def train(self, model, *args, **kwargs):
+        model = RayRemoteModel(model)
+        results = self.executor.execute(
+            lambda trainer: trainer.train(model.load(), *args, **kwargs)
+        )
+        return results[0]
+
+    def train_online(self, model, *args, **kwargs):
+        model = RayRemoteModel(model)
+        results = self.executor.execute(
+            lambda trainer: trainer.train_online(model.load(), *args, **kwargs)
+        )
+        return results[0]
+
+    @property
+    def validation_field(self):
+        return self.executor.execute_single(lambda trainer: trainer.validation_field)
+
+    @property
+    def validation_metric(self):
+        return self.executor.execute_single(lambda trainer: trainer.validation_metric)
+
+    def shutdown(self):
+        self.executor.shutdown()
+
+
+class RayPredictor(BasePredictor):
+    def __init__(self, horovod_kwargs, predictor_kwargs):
+        # TODO ray: investigate using Dask for prediction instead of Horovod
+        setting = RayExecutor.create_settings(timeout_s=30)
+        self.executor = RayExecutor(setting, **{**get_horovod_kwargs(), **horovod_kwargs})
+        self.executor.start(executable_cls=RemotePredictor, executable_kwargs=predictor_kwargs)
+
+    def batch_predict(self, model, *args, **kwargs):
+        model = RayRemoteModel(model)
+        results = self.executor.execute(
+            lambda predictor: predictor.batch_predict(model.load(), *args, **kwargs)
+        )
+        return results[0]
+
+    def batch_evaluation(self, model, *args, **kwargs):
+        model = RayRemoteModel(model)
+        results = self.executor.execute(
+            lambda predictor: predictor.batch_evaluation(model.load(), *args, **kwargs)
+        )
+        return results[0]
+
+    def batch_collect_activations(self, model, *args, **kwargs):
+        model = RayRemoteModel(model)
+        return self.executor.execute_single(
+            lambda predictor: predictor.batch_collect_activations(model.load(), *args, **kwargs)
+        )
+
+    def shutdown(self):
+        self.executor.shutdown()
+
+
+class RayBackend(RemoteTrainingMixin, Backend):
+    def __init__(self, horovod_kwargs=None):
+        super().__init__()
+        self._df_engine = DaskEngine()
```

Comment:
do we need to set the dask scheduler to be ray somewhere?

Reply:
Good catch! This must have gotten dropped during a merge at some point. Fixed.